controllers/propagator/replicatedpolicy_controller.go

```diff
@@ -317,6 +330,158 @@ func (r *ReplicatedPolicyReconciler) Reconcile(ctx context.Context, request ctrl
 	return reconcile.Result{}, watcherErr
 }
 
+// setDBAnnotations sets the parent policy ID on the replicated policy and the policy ID for each policy template.
+// If the DB connection is unavailable, it queues up a reconcile for when the DB becomes available.
+func (r *ReplicatedPolicyReconciler) setDBAnnotations(
+	ctx context.Context,
+	rootPolicy *policiesv1.Policy,
+	replicatedPolicy *policiesv1.Policy,
+) {
+	dbParentPolicy := complianceeventsapi.ParentPolicyFromPolicyObj(rootPolicy)
+
+	var parentPolicyID int32
+	var dbAvailable bool
+	var dbAvailableTried bool
+
+	cachedParentPolicyID, ok := r.parentPolicyToID.Load(dbParentPolicy.Key())
+	if ok {
+		parentPolicyID = cachedParentPolicyID.(int32)
+	} else {
+		dbAvailable = r.ComplianceServerCtx.DB != nil && r.ComplianceServerCtx.DB.PingContext(ctx) == nil
+		dbAvailableTried = true
+
+		if dbAvailable {
+			r.ComplianceServerCtx.Lock.RLock()
+			err := dbParentPolicy.GetOrCreate(ctx, r.ComplianceServerCtx.DB)
+			r.ComplianceServerCtx.Lock.RUnlock()
+
+			if err == nil {
+				parentPolicyID = dbParentPolicy.KeyID
+
+				r.parentPolicyToID.Store(dbParentPolicy.Key(), dbParentPolicy.KeyID)
+			} else {
+				log.Error(
+					err,
+					"Failed to get the database ID of the parent policy",
+					"namespace", replicatedPolicy.Namespace,
+					"name", replicatedPolicy.Name,
+				)
```

Comment:
I was trying to imagine what the rest of the function does after this, because I was trying to think if it could just queue up the thing, and then return early, which might be cleaner/faster?

It seems like it would skip trying to connect to the database again, and just just uses whatever it might already have in its cache. But when its cache is missing something, it deletes that annotation from the policy. So in a case where the cache is empty because the controller just restarted, and it (for some reason) can't connect to the database yet, it seems like it removes all those annotations? Is that right, is that fully intentional?

Reply:
I was originally thinking you can't guarantee accuracy if the DB is down and there is no cache entry, but you're right, we can verify the replicated policy didn't change based on the DB unique constraints and reuse the ID from the replicated policy.

I just added this.